Tests/Common/SimulatedDevice.cs

```diff
@@ -120,7 +118,7 @@ public LoRaPayloadData CreateUnconfirmedDataUpMessage(string data, uint? fcnt =
                 MacMessageType.UnconfirmedDataUp,
                 LoRaDevice.DevAddr.Value,
                 fctrlFlags,
-                fcntBytes,
+                unchecked((ushort)fcnt.Value),
```

Comment:
The data frame cannot accommodate anything but the 16-bit counter or the lower 16 bits of a 32-bit counter if it is supported for the device so there's no choice but to trim here to 16 LSB. The full 32-bit counter is maintained on the server. The overflow check is deliberately muted not because we don't want an overflow error, but precisely because we don't want one if the surrounding code/project (ever) has it in effect.

Reply:
And we could not do it the other way e.g. change the method signature to ushort (since it's creating a data frame I think this would be appropriate) and infer the bigger uint, if the flag `Supports32BitFCnt` is true?

Just curiosity - I think it's acceptable to leave as is since it's test code.

Reply:
> e.g. change the method signature to ushort

The [signature (of the `LoRaPayloadData` constructor) was updated in this PR to take a `ushort`](https://github.com/Azure/iotedge-lorawan-starterkit/pull/1280/files#diff-434a94aed910f6c38befd5bd0e31ceac7a1ea711b9ac654b3a05b0b71d1fc964R151). That's why there is a cast to `ushort` here. Are you talking about the signature of some other method?

> and infer the bigger uint, if the flag `Supports32BitFCnt` is true?

The full 32-bit counter is sent as the last parameter if supported but it's never used or encoded in the data frame. It's just used to initialize the `Server32BitFcnt` property of `LoRaPayloadData`.

Reply:
Sorry, I was talking about this method we are discussing here, `CreateUnconfirmedDataUpMessage`

Reply:
I see. I believe @bastbu is taking over the simulator from @Mandur so perhaps this is something that can be addressed as he starts to look into it more closely?

Reply:
The [cast to `ushort` was there before](https://github.com/Azure/iotedge-lorawan-starterkit/blob/ddd53c77398012704eaf7bf78be250bbccdd9ff9/Tests/Common/SimulatedDevice.cs#L98). What I've added is the _unchecked_ conversion, which has two effects:

- It makes explicit what was the default
- It will maintain the unchecked conversion even if the default is changed for the project or the surrounding code.

As for `CreateUnconfirmedDataUpMessage`, the 16- vs. 32-bit counter support is a run-time (as opposed to compile-time) decision that depends on `Supports32BitFCnt` for a device. Therefore `CreateUnconfirmedDataUpMessage` should remain `fcnt` being typed as `uint`. If anything, we should throw a run-time exception if `fcnt` > 65535 but the device doesn't support a 32-bit **FCnt** and [my suggestion](https://github.com/Azure/iotedge-lorawan-starterkit/pull/1280#discussion_r785040957) was that this could be tackled in a separate PR if necessary. It would just identify that the test setup is incorrect.